Simplify 2D context memory cost calculations

kkinnunen-apple opened the pull request:
#### 6939c8db7b65abc775360f3586781a975024dcd0
<pre>
Simplify 2D context memory cost calculations
<a href="https://bugs.webkit.org/show_bug.cgi?id=273769">https://bugs.webkit.org/show_bug.cgi?id=273769</a>
<a href="https://rdar.apple.com/127601234">rdar://127601234</a>

Reviewed by Simon Fraser.

Avoid calling into ImageBuffer from GC threads. Instead, use the already
existing CanvasBase member variable to keep track of the canvas memory
use.

Set all the canvas memory as &quot;external memory cost&quot;. This is the memory
that is expected to be allocated from other sources than
bmalloc/libcmalloc heap. As 2D context backing store is currently in
GPUP, it cannot be attributed to bmalloc. The Extra/External
calculations are duplicated in other categories, and should be fixed
in other work.

Use the memory cost calculations in OffscreenCanvas, too.

Notify inspector of context changes after all the member state has been
updated. Inspector calls might re-enter calling back any member function
so it&apos;s better that the member state is defined.

Remove unused s_activePixelMemory.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::memoryCost const):
(WebCore::CanvasBase::externalMemoryCost const):
(WebCore::CanvasBase::setImageBuffer const):
(WebCore::CanvasBase::activePixelMemory): Deleted.
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::hasCreatedImageBuffer const):
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::populateBackendInfo):
(WebCore::ImageBuffer::memoryCost const):
(WebCore::ImageBuffer::externalMemoryCost const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::calculateExternalMemoryCost): Deleted.
* Source/WebCore/platform/graphics/NullImageBufferBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::calculateExternalMemoryCost): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::calculateExternalMemoryCost): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:

Canonical link: <a href="https://commits.webkit.org/278797@main">https://commits.webkit.org/278797@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76519e850c4654a34d2b61ad3df49e27fdc437eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51462 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30772 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54729 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2155 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37095 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1835 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41906 "") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28409 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44380 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23030 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25715 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1639 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47684 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1720 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56321 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26581 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49300 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27821 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44442 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48477 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11290 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28714 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27556 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -264,39 +255,31 @@ void CanvasBase::setSize(const IntSize& size)
         m_imageBuffer->context().setStrokeThickness(1);
         m_contextStateSaver = makeUnique<GraphicsContextStateSaver>(m_imageBuffer->context());
-
+    }
```
You don't change `m_size` if `m_imageBuffer` is null. Is that OK?

kkinnunen-apple replied:
Yeah, it's the existing behavior, did not want to change it.
There's some buffer stealing behavior with `setImageBuffer(nullptr)` that is not a resize operation. But future patches should remove those calls, they're a bit hacky. In that work the RefPtr here should be changed to Ref

kkinnunen-apple committed:
Simplify 2D context memory cost calculations
https://bugs.webkit.org/show_bug.cgi?id=273769
rdar://127601234

Reviewed by Simon Fraser.

Avoid calling into ImageBuffer from GC threads. Instead, use the already
existing CanvasBase member variable to keep track of the canvas memory
use.

Set all the canvas memory as "external memory cost". This is the memory
that is expected to be allocated from other sources than
bmalloc/libcmalloc heap. As 2D context backing store is currently in
GPUP, it cannot be attributed to bmalloc. The Extra/External
calculations are duplicated in other categories, and should be fixed
in other work.

Use the memory cost calculations in OffscreenCanvas, too.

Notify inspector of context changes after all the member state has been
updated. Inspector calls might re-enter calling back any member function
so it's better that the member state is defined.

Remove unused s_activePixelMemory.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::memoryCost const):
(WebCore::CanvasBase::externalMemoryCost const):
(WebCore::CanvasBase::setImageBuffer const):
(WebCore::CanvasBase::activePixelMemory): Deleted.
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::hasCreatedImageBuffer const):
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::populateBackendInfo):
(WebCore::ImageBuffer::memoryCost const):
(WebCore::ImageBuffer::externalMemoryCost const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::calculateExternalMemoryCost): Deleted.
* Source/WebCore/platform/graphics/NullImageBufferBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::calculateExternalMemoryCost): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::calculateExternalMemoryCost): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:

Canonical link: https://commits.webkit.org/278797@main
